### PR TITLE
Enhance Test Utility functions

### LINKED ISSUE #12: Enhance testing infrastructure
### Context

Currently, we do assertions of events that happen on the supervision system, and we lock goroutines from the supervised children. There are scenarios where we would like to block once an event happens in the supervision system, and this cannot be possible using the existing test infrastructure.

### Acceptance Criteria

* We have a testing API that allows us to wait for an event to happen
* We don't have to wait for the supervision system to completely stop to do assertion on events that happened
* We can adapt existing tests to new API

### PR DESCRIPTION
**Pull request by roman**
Closes #12

### Context

Currently, we do assertions of events that happen on the supervision system, and we lock goroutines from the supervised children. There are scenarios where we would like to block until as soon as an event happens in the supervision system (not when the supervision system finished), and this cannot be possible using the existing test infrastructure.

### Acceptance Criteria

* [X] We have a testing API that allows us to wait for an event to happen
* [X] We don't have to wait for the supervision system to completely stop to do assertion on events that happened
* [X] We can adapt existing tests to new API

### CODE REVIEW

**Comment by roman** [comment]
Hey @julio-garcia-fc or @Leonid99, if you can take a peek at this, I would highly appreciate it. This is enhancing the test infrastructure

File: stest/evmanager.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	stepFn func(interface{}, s.Event) (bool, interface{}),
+) interface{} {
+	em := ei.evManager
+	shouldContinue := true
+	acc := zero
+
+	for {
+		for {
+			em.evBufferCond.L.Lock()
+			if ei.evIx >= len(*em.evBuffer) && !em.evDone {
```
Comment (julio-garcia-fc) [inline stest/evmanager.go]: This is very hard to follow. Another thread must be updating `ei` and `em` and we must wait here until the event is emitted?

**Comment by roman** [comment]
> It looks like you renamed a file and reorganized code? If so, it might have been easier to follow if I were able to see the diffs with the simple no-op post rename. :man_shrugging:

Sorry, I don't think I'm following, but I want to learn something here (I want to make it easier for the reviewer, in this case you).

What do you mean with "a simple no-op post rename"?

**Comment by julio-garcia-fc** [comment]
> > It looks like you renamed a file and reorganized code? If so, it might have been easier to follow if I were able to see the diffs with the simple no-op post rename. 🤷‍♂
> 
> Sorry, I don't think I'm following, but I want to learn something here (I want to make it easier for the reviewer, in this case you).
> 
> What do you mean with "a simple no-op post rename"?

I was a bit cryptic. Sorry. All I mean is that if you rename the file in one commit (maybe even PR) and then proceed with the other changes on a _separate/second commit_, then it is easy for the reviewer to see the actual changes. Otherwise, the tools are not smart enough to just show the changes and all you see is a massive addition of code (unless I have missed a review setting).

File: stest/assertions.go [hunk 2]
```diff
@@ -43,12 +43,21 @@ func AssertExactMatch(t *testing.T, evs []s.Event, preds []EventP) {
...
-// EventP predicates to a list of supervision system events. The input events
-// need to match in order, but the events do not need to be a 1 to 1 match (e.g.
-// the input events slice length may be bigger than the predicates slice
-// length). This function is useful when we want to test that _some_ events are
-// present in the expected order in a noisy system.
+// verifyPartialMatch is a utility function that matches (in order) a list of
+// EventP predicates to a list of supervision system events.
+//
+// The supervision system events need to match in order all the list of given
+// predicates, however, there has not to be a one to one match between the input
```
Comment (julio-garcia-fc) [inline stest/assertions.go]: Very minor wording/editorial suggestion: s/_there has not to be_/_there does not need to be_/

File: stest/assertions.go [hunk 3]
```diff
@@ -73,11 +82,15 @@ func verifyPartialMatch(preds []EventP, given []s.Event) []EventP {
...
 
 // AssertPartialMatch is an assertion that matches in order a list of EventP
-// predicates to a list of supervision system events. The input events need to
-// match in order, but the events do not need to be a 1 to 1 match (e.g. the
-// input events slice length may be bigger than the predicates slice length).
-// This function is useful when we want to test that _some_ events are present
-// in the expected order in a noisy system.
+// predicates to a list of supervision system events.
+//
+// The input events need to match in the predicate corder, but the events do not
```
Comment (julio-garcia-fc) [inline stest/assertions.go]: s/corder/order/
